File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        //kadane's algorithm: TIme: O(n) and Space:O(1)
        int mx = INT_MIN;
        int sum = 0;
        for(auto it:nums)
        {
            sum += it;
            mx = max(mx,sum);
            if(sum < 0)
                sum = 0;
        }
        return mx;
    }
};

File: 53-maximum-subarray/53-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        // Kadane's algorithm: Time:O(n) and Space:O(1)
        int mx = Integer.MIN_VALUE;
        int sum = 0;
        for(int i= 0;i<nums.length;i++){
            sum += nums[i];
            mx = Math.max(sum,mx);
            if(sum<0)
                sum = 0;
        }
        return mx;
    }
}